tests/handlers/records-subscribe.spec.ts

```diff
@@ -0,0 +1,773 @@
+            protocolRole : 'friend',
+          });
+
+          const subscriptionWithRoleReply = await dwn.processMessage(alice.did, bobSubscriptionWithRole.message, { subscriptionHandler: addRecord });
+          expect(subscriptionWithRoleReply.status.code).to.equal(200);
+          expect(subscriptionWithRoleReply.subscription).to.exist;
+
+
+          // Alice writes three 'chat' records
+          const chatRecordIds = [];
+          for (let i = 0; i < 3; i++) {
+            const chatRecord = await TestDataGenerator.generateRecordsWrite({
+              author       : alice,
+              recipient    : alice.did,
+              protocol     : protocolDefinition.protocol,
+              protocolPath : 'chat',
+              published    : false,
+              data         : new TextEncoder().encode('Bob can read this cuz he is my friend'),
+            });
+            const chatReply = await dwn.processMessage(alice.did, chatRecord.message, { dataStream: chatRecord.dataStream });
+            expect(chatReply.status.code).to.equal(202);
+            chatRecordIds.push(chatRecord.message.recordId);
+          }
+
+          // there should not be any messages in the subscription without a friend role.
+          expect(noRoleRecords.length).to.equal(0);
+
+          // should have all chat messages
+          expect(recordIds).to.have.members(chatRecordIds);
+        });
+
+        it('allows protocol authorized subscriptions', async () => {
+          // scenario: Alice writes some chat messages writes a chat message.
+          //   Bob, having a thread/participant record, can subscribe to the chat.
+
+          const alice = await DidKeyResolver.generate();
+          const bob = await DidKeyResolver.generate();
+
+          const protocolDefinition = threadRoleProtocolDefinition;
+
+          const protocolsConfig = await TestDataGenerator.generateProtocolsConfigure({
+            author: alice,
+            protocolDefinition
+          });
+          const protocolsConfigureReply = await dwn.processMessage(alice.did, protocolsConfig.message);
+          expect(protocolsConfigureReply.status.code).to.equal(202);
+
+
+          // Alice writes a 'thread' record
+          const threadRecord = await TestDataGenerator.generateRecordsWrite({
+            author       : alice,
+            protocol     : protocolDefinition.protocol,
+            protocolPath : 'thread',
+          });
+          const threadRoleReply = await dwn.processMessage(alice.did, threadRecord.message, { dataStream: threadRecord.dataStream });
+          expect(threadRoleReply.status.code).to.equal(202);
+
+          const filter: RecordsFilter = {
+            protocol     : protocolDefinition.protocol,
+            protocolPath : 'thread/chat',
+            contextId    : threadRecord.message.contextId,
+          };
+
+          const noRoleRecords: string[] = [];
+          const addNoRole = async (message: GenericMessage): Promise<void> => {
```

Comment:
Good call on having this check. But this does bring up a point: doesn't this mean we allow anyone to subscribe to "nothing" which essentially is a very simple mechanism for DDOS? I do wonder if it is worth while thinking about prevention/mitigation further, even if outside the scope of this PR.